src/clj/muuntaja/core.clj

```diff
@@ -68,12 +69,22 @@
        :format format}
       e)))
 
-(defn- create-coder [format type spec spec-opts default-charset [p pf]]
-  (let [g (if (vector? spec)
-            (let [[f opts] spec]
-              (f (merge opts spec-opts)))
-            spec)
-        prepare (if (= type ::decode) protocols/as-input-stream identity)]
+(defn- create-coder [format type spec spec-opts default-charset allow-empty-input? [p pf]]
+  (let [decode? (= type ::decode)
+        g (as-> spec $
+
+                ;; duct-style generator
+                (if (vector? $)
+                  (let [[f opts] $]
+                    (f (merge opts spec-opts)))
+                  $)
+
+                ;; optional guard on empty imput
+                (if (and allow-empty-input? decode?)
+                  (fn [^InputStream is charset]
+                    (if (pos? (.available is)) ($ is charset)))
```

Comment:
InputStream javadocs say that:

> A subclass' implementation of this method may choose to throw an
>    * {@link IOException} if this input stream has been closed by
>     * invoking the {@link #close()} method.

I guess we don't have to think about this?